### PR TITLE
Content Manager - Replace `S3Downloader` for `FileDownloader`

### PR DESCRIPTION
**Pull request by tsarquis88**
|Related issue|
|---|
|#20137|

## Description

This pull request introduces a new class called `FileDownloader` that will replace the `S3Downloader` class.

The new class is just the old one but more general, in a meaning that it can download any file from any HTTP URL, not just S3 bucket files.

Change log:
- `S3Downloader` class renamed and generalized.
- Same for UTs.
- `FactoryDownloader` (and its UTs) adapted.
- Remove mentions of `s3FileName` config field.
- Documentation added.

### Test tool

- Raw file download

Config:
```json
{
    "contentSource": "file",
    "compressionType": "raw",
    "versionedContent": "false",
    "deleteDownloadedContent": true,
    "url": "https://raw.githubusercontent.com/wazuh/wazuh/master/README.md",
    "outputFolder": "/tmp/testProvider",
    "dataFormat": "json",
    "contentFileName": "leeme.md",
    "databasePath": "/tmp/content_updater/rocksdb",
    "offset": 0
}
```

Output:
```bash
# ./content_manager_test_tool 
ActionOrchestrator - Starting process
API offset to be used: 0
Output folders created.
FactoryContentUpdater - Starting process
Creating 'file' downloader
Creating 'raw' content decompressor
Version updater not needed
Downloaded content cleaner created
FactoryContentUpdater - Finishing process
ActionOrchestrator - Finishing process
ActionOrchestrator - Running process
FileDownloader - Download done successfully
SkipStep - Executing
PubSubPublisher - Data published
SkipStep - Executing
All files in the folder have been deleted.

changing interval
Action: Initiating scheduling action for test
ActionOrchestrator - Running process
Content file didn't change from last download
FileDownloader - Download done successfully
SkipStep - Executing
PubSubPublisher - No data data to publish
SkipStep - Executing
All files in the folder have been deleted.

Action: Initiating scheduling action for test
ActionOrchestrator - Running process
Content file didn't change from last download
FileDownloader - Download done successfully
SkipStep - Executing
PubSubPublisher - No data data to publish
SkipStep - Executing
All files in the folder have been deleted.


# tree /tmp/testProvider/
/tmp/testProvider/
|-- contents
|   `-- leeme.md
`-- downloads

2 directories, 1 file


# head /tmp/testProvider/contents/leeme.md 
# Wazuh

[![Slack](https://img.shields.io/badge/slack-join-blue.svg)](https://wazuh.com/community/join-us-on-slack/)
[![Email](https://img.shields.io/badge/email-join-blue.svg)](https://groups.google.com/forum/#!forum/wazuh)
[![Documentation](https://img.shields.io/badge/docs-view-green.svg)](https://documentation.wazuh.com)
[![Documentation](https://img.shields.io/badge/web-view-green.svg)](https://wazuh.com)
[![Coverity](https://scan.coverity.com/projects/10992/badge.svg)](https://scan.coverity.com/projects/wazuh-wazuh)
[![Twitter](https://img.shields.io/twitter/follow/wazuh?style=social)](https://twitter.com/wazuh)
[![YouTube](https://img.shields.io/youtube/views/peTSzcAueEc?style=social)](https://www.youtube.com/watch?v=peTSzcAueEc)
```

### Coverage

Test coverage at 95%.

image

### CODE REVIEW

**Review by jftuduri — COMMENTED** [review]
Partial review

**Review by sebasfalcone — COMMENTED** [review]
Partial review

File: src/shared_modules/content_manager/tests/unit/fileDownloader_test.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+{
+    // Set up expected data.
+    nlohmann::json expectedData;
+    expectedData["paths"] = m_spUpdaterContext->data.at("paths");
+    expectedData["stageStatus"] = nlohmann::json::array();
+    expectedData["stageStatus"].push_back(FAIL_STATUS);
+
+    // Set invalid config data. This will make the downloader to download from 'localhost:4444/invalid_file'.
+    m_spUpdaterBaseContext->configData["compressionType"] = "raw";
+    m_spUpdaterBaseContext->configData["url"].get_ref<std::string&>() += "invalid_file";
```
Comment (jftuduri) [inline src/shared_modules/content_manager/tests/unit/fileDownloader_test.cpp:85]: Total nitpick:
How about..
```suggestion
    m_spUpdaterBaseContext->configData["url"] = BASE_URL + "invalid_file";
```
(BASE_URL must be made a std::string)

File: src/shared_modules/content_manager/tests/unit/fileDownloader_test.cpp
Comment (jftuduri) [inline src/shared_modules/content_manager/tests/unit/fileDownloader_test.cpp:1]: +1 good test cases

File: src/shared_modules/content_manager/tests/unit/fileDownloader_test.cpp (hunk 1 above)
Comment (tsarquis88) [reply src/shared_modules/content_manager/tests/unit/fileDownloader_test.cpp]: Done! Much cleaner.

File: src/shared_modules/content_manager/testtool/main.cpp [hunk 2]
```diff
@@ -18,7 +18,7 @@
  * @url: URL where the content is located.
  * @outputFolder: if defined, the content will be downloaded to this folder.
  * @dataFormat: Format of the content downloaded or after decompression.
- * @fileName: Name of the file to download or file name in case of raw content.
+ * @contentFileName: Name for the downloaded file (except if using the offline contentSource).
```
[the remark this answers is not in the document]
Comment (tsarquis88) [reply src/shared_modules/content_manager/testtool/main.cpp]: Done.

File: src/shared_modules/content_manager/tests/unit/fileDownloader_test.cpp [hunk 3]
```diff
@@ -0,0 +1,240 @@
+/*
+ * Wazuh content manager - Unit Tests
+ * Copyright (C) 2015, Wazuh Inc.
+ * Jun 08, 2023.
```
Comment (sebasfalcone) [inline src/shared_modules/content_manager/tests/unit/fileDownloader_test.cpp:4]: Fix date

File: src/shared_modules/content_manager/src/components/fileDownloader.hpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
+/*
+ * Wazuh content manager
+ * Copyright (C) 2015, Wazuh Inc.
+ * May 02, 2023.
```
Comment (sebasfalcone) [inline src/shared_modules/content_manager/src/components/fileDownloader.hpp:4]: Fix date

**Review by sebasfalcone — COMMENTED** [review]
Just some minor details (non code related), this LGTM

**Review by jftuduri — APPROVED** [review]
LGTM

**Review by sebasfalcone — APPROVED** [review]
LGTM!

**Review by jftuduri — APPROVED** [review]

**Review by kevin-mcm — APPROVED** [review]
LGTM!

**Review by Damian-Mangold — APPROVED** [review]
LGTM !

**Pull request merged**